### PR TITLE
Implemented WeatherApp MVP

### PR DESCRIPTION
**Pull request by batchatT**

### CODE REVIEW

File: app/src/main/java/com/example/weatherapp/CommonWeatherInfo.kt [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    private lateinit var commonWeatherViewModel: CommonWeatherInfoViewModel
+
+    private lateinit var adapter : WeatherAdapter
+
+    private lateinit var weatherList: ArrayList<WeatherInfo>
+
+    private lateinit var binding: FragmentCommonWeatherInfoBinding
+
+    override fun onCreateView(
+        inflater: LayoutInflater, container: ViewGroup?,
```
Comment (denisque75) [inline app/src/main/java/com/example/weatherapp/CommonWeatherInfo.kt]: formatting

File: app/src/main/java/com/example/weatherapp/CommonWeatherInfo.kt [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        binding = DataBindingUtil.inflate(
+            inflater,
+            R.layout.fragment_common_weather_info, container, false
+        )
+        adapter = WeatherAdapter(weatherList)
+
+        binding.weatherRecycler.adapter = adapter
+        binding.weatherRecycler.layoutManager = LinearLayoutManager(context)
+
+        binding.lifecycleOwner = viewLifecycleOwner
```
Comment (denisque75) [inline app/src/main/java/com/example/weatherapp/CommonWeatherInfo.kt]: separate logic from onCreateView.  Add new private function "setBinding()"

File: app/src/main/java/com/example/weatherapp/CommonWeatherInfo.kt [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+        Log.d("CommonWeatherInfo", "Checl")
+
+        adapter.setOnWeatherLIstener(object : WeatherAdapter.OnWeatherListener {
```
Comment (denisque75) [inline app/src/main/java/com/example/weatherapp/CommonWeatherInfo.kt]: naming

File: app/src/main/java/com/example/weatherapp/CommonWeatherInfo.kt [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+        })
+
+        return binding.root
+    }
+
+
+    fun filter(text: String?) {
+        val temp: ArrayList<WeatherInfo> = ArrayList<WeatherInfo>()
+        for (weather_item in weatherList) {
```
Comment (denisque75) [inline app/src/main/java/com/example/weatherapp/CommonWeatherInfo.kt]: code style

File: app/src/main/java/com/example/weatherapp/CommonWeatherInfo.kt [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            if (text != null) {
+                if (weather_item.city.contains(text, ignoreCase = true)) {
+                    temp.add(weather_item)
+                }
+            }
+        }
+        adapter.filter(temp)
+    }
+
+    private fun addWeather() {
```
Comment (denisque75) [inline app/src/main/java/com/example/weatherapp/CommonWeatherInfo.kt]: single responsibility

Comment (denisque75) [reply app/src/main/java/com/example/weatherapp/CommonWeatherInfo.kt]: Move it to the Mock Repository. And call it from view model

File: app/src/main/java/com/example/weatherapp/DetailedWeatherInfo.kt [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+import androidx.databinding.DataBindingUtil
+import com.example.weatherapp.databinding.FragmentDetailedWeatherInfoBinding
+import java.lang.StringBuilder
+
+class DetailedWeatherInfo : Fragment() {
+
+    private lateinit var binding: FragmentDetailedWeatherInfoBinding
+
+    override fun onCreateView(
+        inflater: LayoutInflater, container: ViewGroup?,
```
Comment (denisque75) [inline app/src/main/java/com/example/weatherapp/DetailedWeatherInfo.kt]: formatting

File: app/src/main/java/com/example/weatherapp/MainActivity.kt [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+import androidx.appcompat.app.AppCompatActivity
+import android.os.Bundle
+
+class MainActivity : AppCompatActivity() {
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_main)
+        supportActionBar?.hide(
+
```
Comment (denisque75) [inline app/src/main/java/com/example/weatherapp/MainActivity.kt]: formatting

File: app/src/main/java/com/example/weatherapp/StartFragment.kt [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+import android.os.Looper
+import androidx.fragment.app.Fragment
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import androidx.navigation.fragment.findNavController
+
+class StartFragment : Fragment() {
+    override fun onCreateView(
+        inflater: LayoutInflater, container: ViewGroup?,
```
Comment (denisque75) [inline app/src/main/java/com/example/weatherapp/StartFragment.kt]: formatting

File: app/src/main/java/com/example/weatherapp/WeatherAdapter.kt [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+import android.view.ViewGroup
+import android.widget.TextView
+import androidx.recyclerview.widget.RecyclerView
+import kotlin.text.StringBuilder
+
+class WeatherAdapter(weatherList: ArrayList<WeatherInfo>) : RecyclerView.Adapter<WeatherAdapter.ViewHolder>() {
+
+    private lateinit var onWeatherListener : OnWeatherListener
+
+    fun setOnWeatherLIstener(onWeatherListener: OnWeatherListener){
```
Comment (denisque75) [inline app/src/main/java/com/example/weatherapp/WeatherAdapter.kt]: remove it

File: app/src/main/java/com/example/weatherapp/WeatherAdapter.kt [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import android.widget.TextView
+import androidx.recyclerview.widget.RecyclerView
+import kotlin.text.StringBuilder
+
+class WeatherAdapter(weatherList: ArrayList<WeatherInfo>) : RecyclerView.Adapter<WeatherAdapter.ViewHolder>() {
+
+    private lateinit var onWeatherListener : OnWeatherListener
```
Comment (denisque75) [inline app/src/main/java/com/example/weatherapp/WeatherAdapter.kt]: add setter here

File: app/src/main/java/com/example/weatherapp/WeatherAdapter.kt [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+            val temp = StringBuilder(" ")
+            if(item.degree > 0) {
+                temp.append("+")
+            }
+            temp.append(item.degree).append("°C")
+            degree.text = temp.toString()
+        }
+    }
+
+    var data = weatherList
```
Comment (denisque75) [inline app/src/main/java/com/example/weatherapp/WeatherAdapter.kt]: useless var

File: app/src/main/java/com/example/weatherapp/WeatherAdapter.kt [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+
+        init{
+            itemView.setOnClickListener{
+                listener.onWeatherClick(bindingAdapterPosition)
+            }
+        }
+
+        private val city: TextView = itemView.findViewById(R.id.city_text)
+        private val degree: TextView = itemView.findViewById(R.id.degree_text)
+        fun bind (item:WeatherInfo){
```
Comment (denisque75) [inline app/src/main/java/com/example/weatherapp/WeatherAdapter.kt]: fun bind (item : WeatherInfo) {

File: app/src/main/java/com/example/weatherapp/WeatherInfo.kt [hunk 13]
```diff
@@ -0,0 +1,13 @@
+package com.example.weatherapp
+
+import android.os.Parcelable
+import kotlinx.parcelize.Parcelize
+
+@Parcelize
+data class WeatherInfo(
+    val city: String,
+    val degree:Int,
```
Comment (denisque75) [inline app/src/main/java/com/example/weatherapp/WeatherInfo.kt]: degree: Int

File: app/src/main/res/layout/fragment_start.xml [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    tools:context=".StartFragment">
+
+    <TextView
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:text="APP"
```
Comment (denisque75) [inline app/src/main/res/layout/fragment_start.xml]: add string resource

File: app/src/main/res/layout/fragment_start.xml [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+        android:textSize="30sp"
```
Comment (denisque75) [inline app/src/main/res/layout/fragment_start.xml]: dimen resource

File: app/src/main/res/layout/fragment_start.xml [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+        app:layout_constraintEnd_toEndOf="@id/weather_text"
+        app:layout_constraintTop_toBottomOf="@id/weather_text"/>
+
+    <TextView
+        android:id="@+id/weather_text"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:text="@string/weather"
+        android:textColor="@color/black"
+        android:textSize="50sp"
```
Comment (denisque75) [inline app/src/main/res/layout/fragment_start.xml]: dimen

File: app/src/main/res/layout/weather_recycler_item.xml [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tool = "http://schemas.android.com/apk/res-auto">
+
+    <TextView
+        android:id="@+id/city_text"
+        android:layout_width="0dp"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="20dp"
+        android:gravity="start"
+        android:text="Copenhagen"
```
Comment (denisque75) [inline app/src/main/res/layout/weather_recycler_item.xml]: resource

File: app/src/main/res/layout/weather_recycler_item.xml [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+        android:textColor="@color/black"
+        android:textSize="30sp"
```
Comment (denisque75) [inline app/src/main/res/layout/weather_recycler_item.xml]: resource

File: app/src/main/res/navigation/weather_graph.xml [hunk 19]
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="utf-8"?>
+<navigation xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:id="@+id/weather_graph"
+    app:startDestination="@id/startFragment">
+    <fragment
```
Comment (denisque75) [inline app/src/main/res/navigation/weather_graph.xml]: make it as a separate activity

File: app/src/main/java/com/example/weatherapp/commonWeather/CommonWeatherInfo.kt [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+import androidx.recyclerview.widget.LinearLayoutManager
+import com.example.weatherapp.CustomRepository
+import com.example.weatherapp.R
+import com.example.weatherapp.WeatherAdapter
+import com.example.weatherapp.WeatherInfo
+import com.example.weatherapp.databinding.FragmentCommonWeatherInfoBinding
+import java.util.ArrayList
+
+
+class CommonWeatherInfo : Fragment() {
```
Comment (denisque75) [inline app/src/main/java/com/example/weatherapp/commonWeather/CommonWeatherInfo.kt]: name of your components should show what is the component is. For example your fragment should be named CommonWeatherInfoFragment
CommonWeatherInfo view model - CommonWeatherInfoFragmentViewModel etc.

File: app/src/main/java/com/example/weatherapp/CustomRepository.kt [hunk 21]
```diff
@@ -0,0 +1,14 @@
+package com.example.weatherapp
+
+class CustomRepository : RepositoryInterface {
```
Comment (denisque75) [inline app/src/main/java/com/example/weatherapp/CustomRepository.kt]: strange naming

File: app/src/main/java/com/example/weatherapp/RepositoryInterface.kt [hunk 22]
```diff
@@ -0,0 +1,5 @@
+package com.example.weatherapp
+
+interface RepositoryInterface{
```
Comment (denisque75) [inline app/src/main/java/com/example/weatherapp/RepositoryInterface.kt]: strange name

File: app/src/main/java/com/example/weatherapp/DetailedWeatherInfoFragment.kt [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_detailed_weather_info, container, false)
+
+        getBundle()
+
+        return binding.root
+    }
+
+    private fun getBundle(){
+        val bundle = this.arguments
+        if (bundle != null) {
```
Comment (denisque75) [inline app/src/main/java/com/example/weatherapp/DetailedWeatherInfoFragment.kt]: use kotlin features
f.e. use bundle?.let { //your code }

File: app/src/main/java/com/example/weatherapp/DetailedWeatherInfoFragment.kt [hunk 24]
```diff
@@ -0,0 +1,47 @@
...
+            val weather = bundle.getParcelable<WeatherModel>("Weather")
+            val temp = StringBuilder("")
+            if (weather != null) {
```
Comment (denisque75) [inline app/src/main/java/com/example/weatherapp/DetailedWeatherInfoFragment.kt]: use kotlin features
f.e. use weather?.let { //your code }

File: app/src/main/java/com/example/weatherapp/commonWeather/CommonWeatherInfoFragment.kt [hunk 25]
```diff
@@ -0,0 +1,80 @@
...
+        setBinding()
+
+        commonWeatherFragmentViewModel.weatherList.observe(viewLifecycleOwner, {
+            commonWeatherFragmentViewModel.weatherList.value?.let { adapter.setItems(it) }
+        })
+
+        adapter.setOnWeatherClickListener(object: WeatherAdapter.OnWeatherListener {
+            override fun onWeatherClick(position: Int) {
+                val bundle = Bundle()
+                bundle.putParcelable("Weather",
```
Comment (denisque75) [inline app/src/main/java/com/example/weatherapp/commonWeather/CommonWeatherInfoFragment.kt]: use constants

File: app/src/main/java/com/example/weatherapp/commonWeather/WeatherAdapter.kt [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+    class ViewHolder(itemView: View, listener: OnWeatherListener) :
+        RecyclerView.ViewHolder(itemView) {
+
+        init {
+            itemView.setOnClickListener {
+                listener.onWeatherClick(bindingAdapterPosition)
+            }
+        }
+
+        private val city: TextView = itemView.findViewById(R.id.city_text)
```
Comment (denisque75) [inline app/src/main/java/com/example/weatherapp/commonWeather/WeatherAdapter.kt]: don't use FBI

File: app/src/main/java/com/example/weatherapp/commonWeather/WeatherAdapter.kt [hunk 27]
```diff
@@ -0,0 +1,69 @@
...
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import android.widget.TextView
+import androidx.recyclerview.widget.RecyclerView
+import com.example.weatherapp.R
+import com.example.weatherapp.WeatherModel
+import kotlin.text.StringBuilder
+
+class WeatherAdapter() : RecyclerView.Adapter<WeatherAdapter.ViewHolder>() {
```
Comment (denisque75) [inline app/src/main/java/com/example/weatherapp/commonWeather/WeatherAdapter.kt]: class WeatherAdapter(var data: T)

File: app/src/main/java/com/example/weatherapp/commonWeather/WeatherAdapter.kt [hunk 28]
```diff
@@ -0,0 +1,69 @@
...
+
+    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
+
+            val item = data[position]
+            holder.bind(item)
+
+    }
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
+        val view = LayoutInflater.from(parent.context)
```
Comment (denisque75) [inline app/src/main/java/com/example/weatherapp/commonWeather/WeatherAdapter.kt]: use data\view binding

File: app/src/main/java/com/example/weatherapp/network/NetworkRepositoryInterface.kt [hunk 29]
```diff
@@ -0,0 +1,8 @@
+package com.example.weatherapp.network
+
+import androidx.lifecycle.MutableLiveData
+import com.example.weatherapp.WeatherModel
+
+interface NetworkRepositoryInterface{
```
Comment (denisque75) [inline app/src/main/java/com/example/weatherapp/network/NetworkRepositoryInterface.kt]: name it just NetworkRepository

File: app/src/main/java/com/example/weatherapp/network/WeatherNetworkRepository.kt [hunk 30]
```diff
@@ -0,0 +1,73 @@
...
+import androidx.lifecycle.MutableLiveData
+import com.example.weatherapp.WeatherModel
+import retrofit2.Call
+import retrofit2.Callback
+import retrofit2.Response
+import retrofit2.Retrofit
+import retrofit2.converter.gson.GsonConverterFactory
+import kotlin.coroutines.coroutineContext
+
+class WeatherNetworkRepository : NetworkRepositoryInterface {
```
Comment (denisque75) [inline app/src/main/java/com/example/weatherapp/network/WeatherNetworkRepository.kt]: pass the retrofit interface as parameter NetworkApiInterface

File: app/src/main/java/com/example/weatherapp/commonWeather/CommonWeatherInfoViewModel.kt [hunk 31]
```diff
@@ -0,0 +1,18 @@
+package com.example.weatherapp.commonWeather
+
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import com.example.weatherapp.*
+import com.example.weatherapp.network.NetworkRepositoryInterface
+
+class CommonWeatherInfoFragmentViewModel(var networkRepository: NetworkRepositoryInterface) : ViewModel() {
```
Comment (denisque75) [inline app/src/main/java/com/example/weatherapp/commonWeather/CommonWeatherInfoViewModel.kt]: make it val
make it private

File: app/src/main/java/com/example/weatherapp/commonWeather/CommonWeatherInfoViewModel.kt [hunk 32]
```diff
@@ -0,0 +1,18 @@
...
+
+    val weatherList: LiveData<List<WeatherModel>>
+        get() = _weatherList
+    private var _weatherList = networkRepository.getFromNetwork(MutableLiveData(), "")
```
Comment (denisque75) [inline app/src/main/java/com/example/weatherapp/commonWeather/CommonWeatherInfoViewModel.kt]: var _weatherListLiveData = MutableLiveData<List<WeatherModel>>()

File: app/src/main/java/com/example/weatherapp/commonWeather/CommonWeatherInfoViewModel.kt [hunk 33]
```diff
@@ -0,0 +1,18 @@
...
+
+    fun onClickFindWeatherInfo(cityName: String){
+        _weatherList = networkRepository.getFromNetwork(MutableLiveData(), cityName)
```
Comment (denisque75) [inline app/src/main/java/com/example/weatherapp/commonWeather/CommonWeatherInfoViewModel.kt]: use your live data  (which you've already created)

File: app/src/main/java/com/example/weatherapp/commonWeather/CommonWeatherInfoFragment.kt [hunk 34]
```diff
@@ -0,0 +1,80 @@
...
+
+    private lateinit var commonWeatherFragmentViewModel: CommonWeatherInfoFragmentViewModel
+
+    private lateinit var adapter: WeatherAdapter
+
+    private lateinit var binding: FragmentCommonWeatherInfoBinding
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        commonWeatherFragmentViewModel = CommonWeatherInfoFragmentViewModel(WeatherNetworkRepository())
```
Comment (denisque75) [inline app/src/main/java/com/example/weatherapp/commonWeather/CommonWeatherInfoFragment.kt]: use viewmodel factory

File: app/src/main/java/com/example/weatherapp/network/WeatherNetworkRepository.kt [hunk 35]
```diff
@@ -0,0 +1,73 @@
...
+                }
+        }
+        return weatherListLiveData
+    }
+
+    private fun creatingRequest(
+        weatherLiveData: MutableLiveData<WeatherModel>,
+        cityName: String
+    ) {
+        if (cityName.isNotEmpty() && !cityName.all { it in " " }) {
```
Comment (denisque75) [inline app/src/main/java/com/example/weatherapp/network/WeatherNetworkRepository.kt]: use String().trim()

File: app/src/main/java/com/example/weatherapp/network/WeatherNetworkRepository.kt [hunk 36]
```diff
@@ -0,0 +1,73 @@
...
+                    ) {
+                        if (response.isSuccessful && response.body()?.weather?.isEmpty() == false) {
+                            response.body()?.let { weatherLiveData.value = it }
+                            weatherLiveData.value?.let { weatherList.add(it) }
+                            weatherListLiveData.value = weatherList
+                        }
+                    }
+
+                    override fun onFailure(call: Call<WeatherModel>, t: Throwable) {
+                        Log.d("TestRepo", "Failure: $t")
```
Comment (denisque75) [inline app/src/main/java/com/example/weatherapp/network/WeatherNetworkRepository.kt]: handle errors (using live data (for errors))

File: app/src/main/java/com/example/weatherapp/DetailedWeatherInfoFragment.kt (hunk 24 above)
Comment (batchatT) [reply app/src/main/java/com/example/weatherapp/DetailedWeatherInfoFragment.kt]: done

File: app/src/main/java/com/example/weatherapp/commonWeather/CommonWeatherInfoFragment.kt (hunk 25 above)
Comment (batchatT) [reply app/src/main/java/com/example/weatherapp/commonWeather/CommonWeatherInfoFragment.kt]: done

File: app/src/main/java/com/example/weatherapp/commonWeather/WeatherAdapter.kt (hunk 27 above)
Comment (batchatT) [reply app/src/main/java/com/example/weatherapp/commonWeather/WeatherAdapter.kt]: done

File: app/src/main/java/com/example/weatherapp/network/NetworkRepositoryInterface.kt (hunk 29 above)
Comment (batchatT) [reply app/src/main/java/com/example/weatherapp/network/NetworkRepositoryInterface.kt]: done

File: app/src/main/java/com/example/weatherapp/commonWeather/CommonWeatherInfoViewModel.kt (hunk 31 above)
Comment (batchatT) [reply app/src/main/java/com/example/weatherapp/commonWeather/CommonWeatherInfoViewModel.kt]: done

File: app/src/main/java/com/example/weatherapp/network/WeatherNetworkRepository.kt (hunk 35 above)
Comment (batchatT) [reply app/src/main/java/com/example/weatherapp/network/WeatherNetworkRepository.kt]: done

File: app/src/main/java/com/example/weatherapp/network/WeatherNetworkRepository.kt (hunk 30 above)
Comment (batchatT) [reply app/src/main/java/com/example/weatherapp/network/WeatherNetworkRepository.kt:14]: done

File: app/src/main/java/com/example/weatherapp/DetailedWeatherInfoFragment.kt (hunk 23 above)
Comment (batchatT) [reply app/src/main/java/com/example/weatherapp/DetailedWeatherInfoFragment.kt]: done

File: app/src/main/java/com/example/weatherapp/commonWeather/CommonWeatherInfoViewModel.kt (hunk 32 above)
Comment (batchatT) [reply app/src/main/java/com/example/weatherapp/commonWeather/CommonWeatherInfoViewModel.kt]: done

File: app/src/main/java/com/example/weatherapp/commonWeather/CommonWeatherInfoViewModel.kt (hunk 33 above)
Comment (batchatT) [reply app/src/main/java/com/example/weatherapp/commonWeather/CommonWeatherInfoViewModel.kt]: done

File: app/src/main/java/com/example/weatherapp/commonWeather/CommonWeatherInfoFragment.kt (hunk 34 above)
Comment (batchatT) [reply app/src/main/java/com/example/weatherapp/commonWeather/CommonWeatherInfoFragment.kt]: done

**Review by denisque75 — APPROVED** [review]

**Pull request merged**